### PR TITLE
 Sessionsrv migration to api

### PR DESCRIPTION
**Pull request by eeyun**
Woo doggies. This was a massive miserable bag of spaghetti snakes. This is a major simplification and @elliott-davis wrote like 99% of this PR.

Signed-off-by: Ian Henry <ihenry@chef.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-api/src/server/error.rs [hunk 1]
```diff
@@ -45,6 +47,8 @@ pub enum Error {
     CircularDependency(String),
     Connection(conn::ConnErr),
     BadRequest(String),
+    DieselError(diesel::result::Error),
+    GetDbPool(r2d2::Error),
```
Comment (chefsalim) [inline components/builder-api/src/server/error.rs]: Minor - maybe make this error more generic - eg, DbError(r2d2::Error)

File: components/builder-api/src/server/framework/middleware.rs [hunk 2]
```diff
@@ -100,7 +104,7 @@ fn authenticate(req: &HttpRequest<AppState>, token: &str) -> error::Result<Sessi
     // Check for a valid personal access token
     if bldr_core::access_token::is_access_token(token) {
         let mut memcache = req.state().memcache.borrow_mut();
-        match memcache.get_session(token) {
+        match memcache.get_access_token(token) {
```
Comment (chefsalim) [inline components/builder-api/src/server/framework/middleware.rs]: This is excellent.. I think we can make it even simpler.  Just have a single get_session/set_session memcache API.  If the cache hits, great.  If the cache misses, check whether the token is a personal access token, and if so, do the validation according to the None branch below.  If the token is not a personal access token, then return session expired.

**Comment by chefsalim** [comment]
Overall this is some really great work - it's such an improvement over the current sessionsrv.

**Comment by raskchanky** [comment]
This is a total nitpick but I might suggest changing the commit message to remove "WIP" and maybe add some more details about what was done.

File: components/builder-api/src/server/services/memcache.rs [hunk 3]
```diff
@@ -112,10 +112,10 @@ impl MemcacheClient {
         self.reset_namespace(&channel_ns_key(origin, channel));
     }
 
-    pub fn get_session(&mut self, user: &str) -> Option<Session> {
-        trace!("Getting session for user {} from memcached", user);
+    pub fn get_token(&mut self, token: &str) -> Option<Session> {
```
Comment (chefsalim) [inline components/builder-api/src/server/services/memcache.rs]: Nitpicking a bit, but the "get_session" naming seems more accurate, and more consistent with the other function naming here (eg, get/set package)..

**Pull request merged**